src/amoc_coordinator.erl

```diff
@@ -0,0 +1,206 @@
+
+-type coordination_item() :: {NoOfUsers :: pos_integer() | all,
+                              coordination_actions()}.
+
+-type normalized_coordination_item() :: {NoOfUsers :: pos_integer() | all,
+                                         [coordination_action()]}.
+
+-type coordination_plan() :: [coordination_item()] | coordination_item().
+
+%% timeout in seconds
+-type coordination_timeout_in_sec() :: pos_integer() | infinity.
+
+-export_type([coordination_plan/0, normalized_coordination_item/0]).
+
+%%%===================================================================
+%%% Api
+%%%===================================================================
+-spec start(atom(), coordination_plan()) -> ok | error.
+start(Name, CoordinationPlan) ->
+    start(Name, CoordinationPlan, ?DEFAULT_TIMEOUT).
+
+-spec start(atom(), coordination_plan(), coordination_timeout_in_sec()) -> ok | error.
+start(Name, CoordinationPlan, Timeout) when ?is_timeout(Timeout) ->
+    Plan = normalize_coordination_plan(CoordinationPlan),
+    case gen_event:start({local, Name}) of
+        {ok, _} ->
+            %% according to gen_event documentation:
+            %%
+            %%    When the event is received, the event manager calls
+            %%    handle_event(Event, State) for each installed event
+            %%    handler, in the same order as they were added.
+            %%
+            %% in reality the order is reversed, the last added handler
+            %% is executed at first. so to ensure that all the items in
+            %% the plan with NoOfUsers =:= all are executed in the very
+            %% end, we need to add them first.
+            AllItemsHandlers = lists:reverse([Item || {all, _} = Item <- Plan]),
+            [gen_event:add_handler(Name, ?MODULE, Item) || Item <- AllItemsHandlers],
+            [gen_event:add_handler(Name, ?MODULE, Item) || {N, _} = Item <- Plan, is_integer(N)],
+            gen_event:add_handler(Name, ?MODULE, {timeout, Name, Timeout}),
+            ok;
+        {error, _} -> error
+    end.
+
+-spec stop(atom()) -> ok.
+stop(Name) ->
+    gen_event:stop(Name).
+
+-spec add(atom(), any()) -> ok.
+add(Name, Data) ->
+    add(Name, self(), Data).
+
+-spec add(atom(), pid(), any()) -> ok.
+add(Name, Pid, Data) ->
+    gen_event:notify(Name, {coordinate, {Pid, Data}}).
+
+-spec reset(atom()) -> ok.
+reset(Name) ->
+    gen_event:notify(Name, reset_coordinator).
+
+%%%===================================================================
+%%% gen_event callbacks
+%%%===================================================================
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Whenever a new event handler is added to an event manager,
+%% this function is called to initialize the event handler.
+%%
+%% @end
+%%--------------------------------------------------------------------
+-spec(init({timeout, atom(), coordination_timeout_in_sec()} | normalized_coordination_item()) ->
```

Comment:
What is the benefit of using this `spec` syntax over the other syntax used in previous lines? Is it possible to be consistent?